osu.Game/Utils/GeometryUtils.cs

```diff
@@ -218,5 +219,143 @@ private static IEnumerable<Vector2> enumerateStartAndEndPositions(IEnumerable<IH
+        private static (Vector2, float) circleFrom(Vector2 a, Vector2 b, Vector2 c)
+        {
+            if (Precision.AlmostEquals(0, (b.Y - a.Y) * (c.X - a.X) - (b.X - a.X) * (c.Y - a.Y)))
+                return circleFrom(a, b);
+
+            // See: https://en.wikipedia.org/wiki/Circumcircle#Cartesian_coordinates
+            float d = 2 * (a.X * (b - c).Y + b.X * (c - a).Y + c.X * (a - b).Y);
+            float aSq = a.LengthSquared;
+            float bSq = b.LengthSquared;
+            float cSq = c.LengthSquared;
+
+            var centre = new Vector2(
+                aSq * (b - c).Y + bSq * (c - a).Y + cSq * (a - b).Y,
+                aSq * (c - b).X + bSq * (a - c).X + cSq * (b - a).X) / d;
+
+            return (centre, Vector2.Distance(a, centre));
+        }
+
+        // Function to return the smallest circle that intersects 2 points
+        private static (Vector2, float) circleFrom(Vector2 a, Vector2 b)
+        {
+            var centre = (a + b) / 2.0f;
+            return (centre, Vector2.Distance(a, b) / 2.0f);
+        }
+
+        // Function to check whether a circle encloses the given points
+        private static bool isValidCircle((Vector2, float) c, ReadOnlySpan<Vector2> points)
+        {
+            // Iterating through all the points to check whether the points lie inside the circle or not
+            foreach (Vector2 p in points)
+            {
+                if (!isInside(c, p)) return false;
+            }
+
+            return true;
+        }
+
+        // Function to return the minimum enclosing circle for N <= 3
+        private static (Vector2, float) minCircleTrivial(ReadOnlySpan<Vector2> points)
+        {
+            if (points.Length > 3)
+                throw new ArgumentException("Number of points must be at most 3", nameof(points));
+
+            switch (points.Length)
+            {
+                case 0:
+                    return (new Vector2(0, 0), 0);
+
+                case 1:
+                    return (points[0], 0);
+
+                case 2:
+                    return circleFrom(points[0], points[1]);
+            }
+
+            // To check if MEC can be determined by 2 points only
+            for (int i = 0; i < 3; i++)
+            {
+                for (int j = i + 1; j < 3; j++)
+                {
+                    var c = circleFrom(points[i], points[j]);
+
+                    if (isValidCircle(c, points))
+                        return c;
+                }
+            }
+
+            return circleFrom(points[0], points[1], points[2]);
+        }
+
+        // Returns the MEC using Welzl's algorithm
+        // Takes a set of input points P and a set R
+        // points on the circle boundary.
+        // n represents the number of points in P that are not yet processed.
+        private static (Vector2, float) welzlHelper(List<Vector2> points, ReadOnlySpan<Vector2> r, int n)
+        {
+            const int max_depth = 4000;
```

Comment:
I am _very_ uncomfortable with this sort of thing. Not only there is _zero_ guarantee that this stack depth is going to be allowed (especially in environments such as mobile), this also seems to arbitrarily perturb the correctness of the algorithm in a way that has not been discussed. I could imagine this could return completely bogus results if one produced an example degenerate enough.

I think an alternative algorithm, or an alternative implementation of *this* one that does not feature recursion, should be considered.

Reply:
I've rewritten the algorithm to use no recursion. It's significantly slower and more memory intensive than the recursive version, probably because using a `Stack` is more expensive than the actual stack, but it's good enough and handles big N.
```
| Method                 | N     | Mean        | Error     | StdDev    | Gen0     | Gen1     | Gen2     | Allocated |
|----------------------- |------ |------------:|----------:|----------:|---------:|---------:|---------:|----------:|
| MinimumEnclosingCircle | 100   |    14.69 us |  0.171 us |  0.160 us |   0.6104 |        - |        - |   5.02 KB |
| MinimumEnclosingCircle | 1000  |   131.40 us |  1.800 us |  1.596 us |   4.8828 |        - |        - |  40.13 KB |
| MinimumEnclosingCircle | 2000  |   273.15 us |  4.784 us |  4.475 us |   9.7656 |   0.9766 |        - |  79.96 KB |
| MinimumEnclosingCircle | 4000  |   483.02 us |  8.360 us |  7.820 us |  19.0430 |   3.9063 |        - | 159.61 KB |
| MinimumEnclosingCircle | 8000  | 1,185.40 us | 23.026 us | 21.539 us |  41.0156 |  41.0156 |  41.0156 |  318.9 KB |
| MinimumEnclosingCircle | 10000 | 1,610.47 us | 30.441 us | 32.572 us | 123.0469 | 123.0469 | 123.0469 | 590.57 KB |
```